compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -1391,25 +1391,24 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
   }
   SmallVector<int64_t> distTileSizes =
       getDefaultDistributedLevelTileSizes(op, distConfig);
-  SmallVector<int64_t> workload(op.getSourceType().getShape());
-  reduceDistributionWorkgroups(workload, distTileSizes,
-                               /*maxTileSizes=*/std::nullopt,
-                               distConfig.vectorSizeHints);
```

Comment:
This is intended because of https://github.com/openxla/iree/issues/15349#issuecomment-1790068368

The reduceDistributionWorkgroups  method used in `getDefaultDistributedLevelTileSizes` set maxTileSizes to 64. So they are limit to `64`. In pack op case, I think we can ignore the `maxTileSize` because they are just transpose or pure data movement. What is your insight about balancing (1) divide work evenly and (2) avoid small dispatches?

After doing many distribution cleanups, I've learned more about it. I had a wrong fix (or maybe it is not feasible at that moment). If we want to model it better, we should just set MaxTileSizes to a very large number. In this context, we don't need to call the method again.

Reply:
I think we should try to achieve similar results more directly, using less steps. It's hard to follow what is going on when we go into this relatively complex reduction function twice. Some of these results are also a bit surprising as for some cases we end up using the whole dimension size (i.e., not tiling) and for other similar cases we end up with a significantly smaller tile size.

Why do we pass `64` in `maxTileSizes` in the first call? Would it make sense to remove it?

> What is your insight about balancing (1) divide work evenly and (2) avoid small dispatches?

(1) was the main motivation for this change as we should make sure the work divides evenly when we are dealing with i4 elements. This, of course, doesn't mean that we have to change the behavior for all the non-i4 cases but we should find a good compromise without increasing the complexity.

(Hopefully you can see how useful the pattern proposal would be for this case, as we could have a well encapsulated heuristic with high priority that matches the i4 cases, while keeping the default heuristics simple and more generic :))

Reply:
> Why do we pass 64 in maxTileSizes in the first call? Would it make sense to remove it?

That is the default when maxTileSizes is not set.

> (1) was the main motivation for this change as we should make sure the work divides evenly when we are dealing with i4 elements. This, of course, doesn't mean that we have to change the behavior for all the non-i4 cases but we should find a good compromise without increasing the complexity.

I think it is not just an i4 issue. If you look at this lit test:

```
%pack = tensor.pack %2 outer_dims_perm = [1, 0] inner_dims_pos = [1, 0] inner_tiles = [16, 1] into %3 : tensor<1200x500000xf32> -> tensor<31250x1200x16x1xf32>

...

// CHECK-DAG: #[[CONFIG:.+]] = #iree_codegen.lowering_config<tile_sizes = {{\[}}[50, 48], [1, 1]]>
```

The pack op has 1200x500000 elements. However, there are `(1200/50) * (1200x500000/48) = 3*10^8` kernel launches. In this case, we are not dividing work evenly?

Reply:
> (1) was the main motivation for this change 

Sorry, I meant it was my main motivation to look into this and post this PR. If we choose a tile size that is not even we would end up with a dynamic dimension somewhere in the memref subviews and we can't currently handle dynamic dimensions in sub-byte buffers. As stated, we shouldn't change the behavior for all the non-i4 cases. Happy to get some insights from you on things that we can try :)

Reply:
> I see, it sounds like a specific issue for i4 types? Is it because we are not able to flatten memref.subview with dynamic dimensions, and it causes issue in emulation?

Subviews shouldnt survive till flattening (you can't flatten subviews in general, what is implemented today is just for testing). Haven't followed all the discussions, but if we are bottoming out on this then something went wrong.

Reply:
Thanks, I gave your suggestions a try. Hopefully the numbers make more sense. No, this is not fixing the `memref.subview` problem we talked about, it's still there. The main motivation is to simplify this logic and minimize dynamic shapes due to distribution sizes.

Reply:
If this is not the memref issue, I think we can ignore my (1) suggestion. We don't want to take the step back. It could make distribution tile sizes not be multiple of the vector length. In this context we are penalizing all the threads. The tile sizes look okay to me. Hopefully we can still get reasonable tile sizes if we flip `distConfig.allowIncompleteTile`  back to true.

Reply:
Let me make it true only for sub-byte types. As mentioned, we can't generate question marks for sub-byte types.

Reply:
I enabled partial tiles for all the cases. Let's try the new logic first and decided later what to do with partial tiles